### ISSUE TITLE
Prompt weighting is counterintuitive/broken

### ISSUE THREAD

**Issue by damian0815** [invoke-ai/InvokeAI#917]
**Describe your environment**
- GPU: mps and cuda (same problem on both)
- VRAM: 16GB / 3GB 
- CPU arch: arm / x64
- OS: macOS / Win10
- Python: conda 
- Branch: development
- Commit: 2cbf1e6f4bc7176c010cfb0e613edc5fa2d88d8c

**Describe the bug**
In brief: prompt weighting doesn't seem to be working, or if it is working it is working in a way that is counterintuitive. 

Splitting up the query into weighted sections and giving each a weight of `1` intuitively should give me back the same output as unweighted, or at least very similar, as if i'd put commas between the subprompts. But in practise `wordA wordB` and `wordA:1 wordB:1` give radically different outputs, often with at least one of the terms no longer tangibly present in the output.

If I enter `mountain man` with a seed of `3254369738` I get this jolly fellow:
```
dream> "mountain man" -s10 -S3254369738 -W512 -H384 -A"k_euler"
```
![000059 3254369738](https://user-images.githubusercontent.com/144366/193883578-d29f4816-c544-48d9-bf62-654f1b3e6100.png)

But I want more mountain, less man. Intuition says I need to weight "man" down and/or "mountain" up, so as a step 0 I break the query up into fragments each with weight 1:

```
dream> "mountain:1 man:1" -s10 -S3254369738 -W512 -H384 -A"k_euler"
```

![000060 3254369738](https://user-images.githubusercontent.com/144366/193883973-563e2d76-1d8c-49cc-b585-0b033b869e34.png)
... where'd my man go?? 

Giving him more weight does not bring him back:
```
dream> "mountain:1 man:2" -s10 -S3254369738 -W512 -H384 -A"k_euler"
```
![000632 3254369738](https://user-images.githubusercontent.com/144366/193910148-01b55aac-dadf-4ce5-94c2-de63437e593b.png)

### To reproduce 

1. Launch `server.py` or `dream.py`
2. Enter a prompt with two or more simple terms and note the relationship between the prompt and the image. For example, with `"big giant head" -S3254369738`, note that the output is indeed a big giant head (see below).
3. Insert commas between the words and observe the output. Note that the image is different, but the relationships are comparably strong (see below).
4. Delete the commas and add a weight of `:1` after each word and observe the output (see below). 
**Expected result:** The image may be different but the relationship between the prompt and the image is comparably strong.
**Actual result:** There is no longer an obvious relationship between the words and image.

"big giant head"
![000616 3254369738](https://user-images.githubusercontent.com/144366/193894918-1254fdcf-d6fa-4ba5-a2a6-730f3f94e26a.png)

"big, giant, head"
![000618 3254369738](https://user-images.githubusercontent.com/144366/193894947-d61261f7-7829-42bf-9e3f-a64e1702f11b.png)

(Different output is expected: commas have a unique CLIP token, which you can verify this by putting a breakpoint in `FrozenCLIPEmbedder.forward` in `ldm/modules/encoders/modules.py` line 440.)

"big:1 giant:1 head:1" 
![000617 3254369738](https://user-images.githubusercontent.com/144366/193894972-cb9f97c0-c247-4d04-92f3-a71934fe8319.png)

The prompt is obviously good in the other images - why is it doing so badly here?

### More examples

This counterintuitive behaviour is reproducible across different subprompt counts. Here's a "green cat":

![000625 3254369738](https://user-images.githubusercontent.com/144366/193896786-be0a8f3d-1498-4b48-bb5b-645e6209a6d9.png)

Perfect. And here's a "green, cat":

![000626 3254369738](https://user-images.githubusercontent.com/144366/193896830-1a28cc3b-c774-4390-9d01-d57b3a1f7d20.png)

Green-tinted overall, but still clearly a green cat. And now for a "green:1 cat:1"

![000624 3254369738](https://user-images.githubusercontent.com/144366/193896956-f120c068-31ea-40c9-aecb-04db3fd5a3d5.png)

It's green but is only vaguely cat-like.

Making the prompt more specific unfortunately doesn't help much:

"tall thin man picking apricots from a tree"
![000607 3254369738](https://user-images.githubusercontent.com/144366/193888664-6caec3ca-e29e-41bf-ae5b-3c7dbe8574f7.png)

"tall thin man:1 picking apricots from a tree:1"
![000608 3254369738](https://user-images.githubusercontent.com/144366/193888745-2e5c6390-9143-4923-89c0-910ce864734a.png)
(where'd the apricots go?)

"tall thin man:1 picking apricots:1 from a tree:1"
![000609 3254369738](https://user-images.githubusercontent.com/144366/193888781-fcc7b50e-63ca-470a-abfa-956a4b836bf8.png)
(where'd the man *and* the apricots go??)

Note that inserting commas produces a different result again:

"tall thin man, picking apricots, from a tree"
![000610 3254369738](https://user-images.githubusercontent.com/144366/193889180-c987fe7a-df65-4a95-89a9-409e197e237b.png)

Unfortunately adding weights to the comma-separated groups makes the man and the apricots disappear again:

"tall thin man:1 , picking apricots:1 , from a tree:1"
![000614 3254369738](https://user-images.githubusercontent.com/144366/193890274-01ed6f87-9d1d-4d97-b418-11cf978290d2.png)

"tall thin man,:1 picking apricots,:1 from a tree:1"
![000615 3254369738](https://user-images.githubusercontent.com/144366/193890509-f026e495-38b4-4b3f-9315-3d0806c59543.png)

**Comment by damian0815** [comment]
I've done a dig through the code and compared it with the implementation in stable-diffusion-webui and I think the cause is an assumption made in `get_uc_and_c`. Namely, this logic here:

```
        c = torch.zeros_like(uc)
        # normalize each "sub prompt" and add it
        for subprompt, weight in weighted_subprompts:
            log_tokenization(subprompt, model, log_tokens, weight)
            c = torch.add(
                c,
                model.get_learned_conditioning([subprompt]),
                alpha=weight,
            )
```
assumes that subprompts should be combined by doing a weighted sum of the individual sub-prompts feature tensors. Studying the outputs and the code from `stable-diffusion-webui` I think this assumption is incorrect. `model.get_learned_conditioning` returns a tensor of size `(1, 77, 768)`: each CLIP token embedding is 768 floats, and there are a maximum of 77 tokens (75 + start token + end token) in a prompt conditioning vector. 

From what I can tell, therefore, the strategy of adding outputs from `get_learned_conditioning` ends up mushing the start tokens together. Instead what needs to happen is that the final content of `c` needs to be made by *appending* each 768-item CLIP token embeddings for each token, multiplied each of them by the weight that token should receive, and then (if `stable-diffusion-webui` is doing things correctly) `c` needs to be finally re-normalised to the same value it had before multiplying through by each token's weight.

This means that if the query is `green:0.7 cat:0.3`, we can't simply call `get_learned_conditioning`. Instead we have to access the clip tokenizer directly and manually tokenize each subprompt, to make a 77-token long list of: `[49406,  1901,  2368, (endid) (endid) (endid) ..... (endid)]` where `1901` is the tokenid for "green" and `2368` is the tokenid for "cat". This would then get passed to the `FrozenCLIPEmbedder.transformer()` to produce a `z` of size `(1, 77, 768)`. Each of the last dimension then gets multiplied by the appropriate weight, so for `green:0.7 cat:0.3` we'd have to do something like this:
```
original_mean = z.mean()
# z[0][0] is (startid)
z[0][1] *= 0.7
z[0][2] *= 0.3
# z[0][3] is (endid)
new_mean = z.mean()
z *= original_mean/new_mean
```
(ref https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/88ec0cf5571883d84abd09196652b3679e359f2e/modules/sd_hijack.py#L271)

This might be some heavy engineering. The cleanest way to do it is to make a new `WeightedFrozenCLIPEmbedder` that inherits from `FrozenCLIPEmbedder` and pass `weights` via `kwargs`. I should be able to implement what is required and prepare a pull request - but I wanted to hold off doing that until the issue could be discussed properly.

**Comment by Any-Winter-4079** [comment]
Unrelated, but I'm realising there has been some (major?) change in the behavior of the samplers recently (?).
`"mountain man" -s10 -S3254369738 -W512 -H384 -A"k_euler"`
Yours
<img width="508" alt="Screenshot 2022-10-04 at 20 58 58" src="https://user-images.githubusercontent.com/50542132/193903347-72219a80-6a98-4b12-ba2e-8c40afaa7e21.png">
Mine (just pulled latest changes from dev)
<img width="516" alt="Screenshot 2022-10-04 at 21 03 33" src="https://user-images.githubusercontent.com/50542132/193904130-01baf15e-d14b-489b-97b2-810c8514fa90.png">

I'm on Mac too.

**Comment by lstein** [comment]
Your analysis sounds convincing. The subprompt weighting work was done in the first weeks of the project and hasn’t been revisited. I’ve been puzzled by the behaviour of weighting and this would explain it. 

Please create a PR if you can. Unfortunately this might not be in time to make it into the impending release.

**Comment by bmaltais** [comment]
This is how it should be. Maybe the name is wrong. It should be called "prompt weighting". Think of it as each prompt independently being weighted. And Each will influence the final image based on the weight it has. It is king of a prompt merge...

You are not the 1st to be stumped by this... but this is a nice feature and should not be altered.

**Comment by Any-Winter-4079** [comment]
I think what @damian0815 says is what most of us think of when presented with prompt weighting. At least that's how I tried to use it (and from an end user's perspective, I think it makes sense).

I'm not saying necessarily replace the current behavior, but at least we should try to additionally add the expected behavior.
@bmaltais Also, if you have a guide on how the current prompt weighting works, it would be appreciated.

**Comment by bmaltais** [comment]
> I think what @damian0815 says is what most of us think of when presented with prompt weighting. At least that's how I tried to use it (and from an end user's perspective, I think it makes sense).
> 
> I'm not saying necessarily replace the current behavior, but at least we should try to additionally add the expected behavior. @bmaltais Also, if you have a guide on how the current prompt weighting works, it would be appreciated.

I don't have a specific guide. I did create a HOWTO to use it as part of crafting outputs by combining more than one prompt... I guess it sort of constitute a poor man guide.

**Comment by Any-Winter-4079** [comment]
Yeah, I checked the guide but I'm not sure I see what I would expect from prompt weighting as an end user (plus there's so many parts, in terms of variations etc. that it's hard for me to know the effect of each aspect).
Here the hair is red:

<img width="300" alt="image" src="https://user-images.githubusercontent.com/50542132/193969317-d8c3a0f7-64a5-4fc5-a902-e17e93a795b2.png">

And here she's a robot
<img width="297" alt="image" src="https://user-images.githubusercontent.com/50542132/193969413-93e86c82-ee13-40e2-9c62-e68aae5a2b0a.png">


And the output is her as a robot with pale face (good!) but mostly blonde hair (bad!). For example, I would expect the hair to be red.
<img width="301" alt="image" src="https://user-images.githubusercontent.com/50542132/193969486-9321e349-51cb-4bf4-8cac-64f1572fe3e1.png">

This seems to give a kind of random mix of the prompts, but I think people expect something more along the lines of
`amber heard as a robot :2`
`amber heard as a robot :3 `-> more robot, less human
`amber heard as a robot :4 `-> even more robot, even less human

**Comment by lstein** [comment]
Here's my poor man's way of prompt weighting:

`amber heard as a robot, scifi illustration`

<img src="https://user-images.githubusercontent.com/111189/193977023-df34eaf2-f805-490a-8978-8c806ca9b194.png" width=256>

`amber heard as a robot robot, scifi illustration`
<img src="https://user-images.githubusercontent.com/111189/193977080-79d58cfa-8afb-4f2d-8c52-06af30250e19.png" width=256>

`amber heard as a robot robot robot, scifi illustration`
<img src="https://user-images.githubusercontent.com/111189/193977105-7e6da098-e011-4218-8593-adf2a2157bcd.png" width=256>

@bmaltais I would never take away the prompt weighting! But I would like to have an intuitive syntax that allows us to emphasize certain concepts and deemphasize others. We already use `[excluded term]` to "forget" a concept, and it works quite well. Try the prompt `photograph of a hamburger` vs `photograph of a hamburger [tomatoes]`. What would be great is to selectively emphasize phrases using (). 

~~~
photograph of a hamburger (cheese) [tomatoes]                               -- more cheese, hold the tomatoes
photograph of a hamburger ((pickles) and cheese) [tomatoes]           -- more condiments - I particularly want pickles
~~~

**Comment by damian0815** [comment]
> This is how it should be. Maybe the name is wrong. It should probably be called "prompt weighting". Think of it as each prompt (up to 75 tokens) independently being weighted. And each will influence the final image based on the weight it has. It is kind of a prompt output kind of merge...

However, it is a major problem that I and evidently numerous other people are stumped by this feature. IMO it should be renamed to prompt *interpolating*, because the weights are being used to interpolate between the sub-prompts, not to add emphasis to them:

![cat-to-green-small](https://user-images.githubusercontent.com/144366/194003828-d3b60e86-f018-40c7-9d3c-8b9ed1f8dfdf.gif)
(this is the result of generating "green:0.01 cat:0.09" through "green:0.99 cat:0.01").

This is indeed a nice feature to have, but it is presented in a way that is surprising, and in UX terms surprise is, basically, bad. It has taken me days of frustrated front-end usage, and several hours of close code analysis, to figure out that it the weights are for *interpolating* prompts, not for *emphasising* prompts. The surprise for me - and the frustration - comes when I change "mountain man" to "mountain:1 man:1" and suddenly SD is not even trying to generate either "mountain" nor "man" but is instead searching for something quite different. It's like if you told your travel agent "I want to go London and Vienna" and they book you two hotels in London and Vienna with a tour bus between, but then when you say "I want to go to London for 1 week and Vienna for 1 week" they cancel the tour bus and book you a hotel for 2 weeks somewhere in the German countryside, exactly halfway between London and Vienna. I am certain that people will have been driven away from InvokeAI's SD and toward Automatic1111's SD because of this surprise.

Again: nice feature, but a niche use-case that, for the say 95% of users who are not interested in abstract art or weird artefacts but just want a way to tweak a prompt they have already spent some time fine-tuning, not one that should be so prominently displayed as "weighting" when that term can ambiguously mean "interpolation weighting" and "attention weighting". I think what would have helped me be less confused is if the syntax forced me to write weights that summed to 0, so that it was more clear that it was a lerp and not a "weighting" that was being applied. At least, it should be renamed to "prompt interpolation" to avoid confusion.

**Comment by damian0815** [comment]
What I'll do is the following. I'll implement the feature of "attention weighting" but I'll do it in a way that isn't bound to this syntax. What needs to be decided is whether 
1. the existing syntax stays and gets renamed "interpolation weighting" and we come up with a new syntax for "attention weighting", or
2. the existing syntax is repurposed to do "attention weighting" and we come up with a new syntax for "interpolation weighting"
